comp/core/workloadmeta/collectors/internal/kubeapiserver/kubeapiserver.go

```diff
@@ -146,6 +153,34 @@ func resourcesWithExplicitMetadataCollectionEnabled(cfg config.Reader) []string
 	return resources
 }
 
+// resourcesForAPMConfig returns the list of resources to collect metadata from
+// for the auto instrumentation configuration. Namespaces are collected in order
+// to utilize namespace labels for target based configuration.
+func resourcesForAPMConfig(cfg config.Reader) []string {
+	// If APM is not enabled, we don't need to collect any resources for the
+	// auto instrumentation configuration.
+	apmEnabled := cfg.GetBool("apm_config.instrumentation.enabled")
+	if !apmEnabled {
+		return nil
+	}
+
+	// Targets is a custom struct type, so we unmarshal it into an interface
+	// slice to avoid the import while still being able to check if it's empty.
+	targets := []interface{}{}
+	err := cfg.UnmarshalKey("apm_config.instrumentation.targets", &targets)
```

Comment:
Shouldn't we be more specific here and check for the namespace selector that needs the namespace labels?

Reply:
In order to support that, `kubeapiserver` would need to depend on `autoinstrumentation` which felt wrong as an import. It also felt a bit more complicated to describe to customers and test. But if you think it's worth it as a trade off, I'm happy to add the import so we can parse the target object.

Reply:
Importing extra dependencies can be problematic, depending on which agent requires them and whether they increase the binary size. We have CI checks in place for this.

Fixing this might not be worth it, but I think that at least we should add a comment here indicating that for simplicity, we collect namespace metadata if targets are defined, even if those targets only refer to pods rather than namespaces. This is something that can be optimized later if needed.